packages/flutter_tools/test/integration.shard/test_driver.dart

```diff
@@ -248,44 +250,71 @@ abstract class FlutterTestDriver {
     );
   }
 
-  // This method isn't racy. If the isolate is already paused,
-  // it will immediately return.
-  Future<Isolate> waitForPause() => waitForDebugEvent('Pause');
-  Future<Isolate> waitForResume() => waitForDebugEvent('Resume');
+  Future<Event> subscribeToPauseEvent(String isolateId) => subscribeToDebugEvent('Pause', isolateId);
+  Future<Event> subscribeToResumeEvent(String isolateId) => subscribeToDebugEvent('Resume', isolateId);
+
+  Future<Isolate> waitForPauseEvent(String isolateId, Future<Event> event) =>
+      waitForDebugEvent('Pause', isolateId, event);
+  Future<Isolate> waitForResumeEvent(String isolateId, Future<Event> event) =>
+      waitForDebugEvent('Resume', isolateId, event);
+
+  Future<Isolate> waitForPause() async => subscribeAndWaitForDebugEvent('Pause', await _getFlutterIsolateId());
+  Future<Isolate> waitForResume() async => subscribeAndWaitForDebugEvent('Resume', await _getFlutterIsolateId());
+
+
+  Future<Isolate> subscribeAndWaitForDebugEvent(String kind, String isolateId) {
+    final Future<Event> event = subscribeToDebugEvent(kind, isolateId);
+    return waitForDebugEvent(kind, isolateId, event);
+  }
+
+  /// Subscribes to debug events containing [kind] and returns a future that
+  /// completes when the [kind] event is received.
+  ///
+  /// Note that this method should be called before the command that triggers
+  /// the event to subscribe to the event in time, for example:
+  ///
+  /// ```
+  ///  var event = subscribeToDebugEvent('Pause', id); // Subscribe to 'pause' events.
+  ///  ...                                             // Code that pauses the app,
+  ///  await waitForDebugEvent('Pause', id, event);    // Isolate is paused now.
+  /// ```
+  Future<Event> subscribeToDebugEvent(String kind, String isolateId) async {
+    final Completer<Event> pauseEvent = Completer<Event>();
+    _debugPrint('Start listening for $kind events');
+
+    final StreamSubscription<Event> pauseSubscription = _vmService.onDebugEvent
+        .where((Event event) {
+          return event.isolate.id == isolateId
+              && event.kind.startsWith(kind);
+        })
+        .listen((Event event) {
+          if (!pauseEvent.isCompleted) {
+            pauseEvent.complete(event);
+          }
+        });
+
```

Comment:
Does this work?

```suggestion
  _vmService.onDebugEvent.where(...).first;
  ```

Reply:
Changed, seems to work on my machine, thanks! Will see what the CI says:)